### PR TITLE
Added partially_hide_scoreboard

### PR DESCRIPTION
**Pull request by jtyliu**
Added an option for the scoreboard to hide before and during a contest and show after the user's window ends.

It should work.

### CODE REVIEW

File: judge/admin/contest.py [hunk 1]
```diff
@@ -111,8 +111,9 @@ class ContestAdmin(VersionAdmin):
     fieldsets = (
         (None, {'fields': ('key', 'name', 'organizers')}),
         (_('Settings'), {'fields': ('is_visible', 'is_external', 'is_virtualable', 'use_clarifications',
-                                    'hide_problem_tags', 'freeze_submissions', 'hide_scoreboard',
-                                    'permanently_hide_scoreboard', 'run_pretests_only', 'access_code')}),
+                                    'hide_problem_tags', 'freeze_submissions', 'partially_hide_scoreboard',
+                                    'hide_scoreboard', 'permanently_hide_scoreboard', 'run_pretests_only',
```
Comment (Ninjaclasher) [inline judge/admin/contest.py]: Swap `partially_hide_scoreboard` and `hide_scoreboard`

File: judge/models/contest.py [hunk 2]
```diff
@@ -81,6 +81,10 @@ class Contest(models.Model):
     registration_end_time = models.DateTimeField(verbose_name=_('registration end time'),
                                                  help_text=_('Allow registration until the specified time.'),
                                                  blank=True, null=True)
+    partially_hide_scoreboard = models.BooleanField(verbose_name=_('Partially hide scoreboard'),
+                                                    help_text=_('Whether the scoreboard should be shown hidden '
+                                                    'until the member\'s contest window is over.'),
```
Comment (Ninjaclasher) [inline judge/models/contest.py]: use `"` around the string to remove the need of `\'`

Comment (Ninjaclasher) [inline judge/models/contest.py]: indent to match the start of the previous quote (see other help texts for reference)

File: judge/models/contest.py [hunk 3]
```diff
@@ -236,6 +255,7 @@ def time_before_start(self):
 
     @property
     def time_before_end(self):
+        # In terms on contest time itself
```
Comment (Ninjaclasher) [inline judge/models/contest.py]: We don't want to blowup the MCPT/WLMOJ to DMOJ/online-judge changes with comments that DMOJ does not have, so please remove these (same with some of the other comments). Feel free to PR these to DMOJ though.

File: judge/models/contest.py [hunk 4]
```diff
@@ -182,17 +186,32 @@ def can_see_scoreboard(self, user):
...
             return False
         if not self.show_scoreboard:
             return False
+        if self.partially_hide_scoreboard and not self.is_finished_contest(user):
+            return False
         return True
 
+    def is_finished_contest(self, user):
+        if user.is_authenticated:
+            profile = user.profile
```
Comment (Ninjaclasher) [inline judge/models/contest.py]: inline this

File: judge/models/contest.py [hunk 5]
```diff
@@ -182,17 +186,32 @@ def can_see_scoreboard(self, user):
...
+            if profile:
```
Comment (Ninjaclasher) [inline judge/models/contest.py]: is there a case where both the user is authenticated and the profile is `None`?

File: judge/models/contest.py [hunk 6]
```diff
@@ -182,17 +186,32 @@ def can_see_scoreboard(self, user):
...
+                for participation in ContestParticipation.objects.filter(virtual=0, user=profile) \
```
Comment (Ninjaclasher) [inline judge/models/contest.py]: This is an O(N) operation that can be done in O(1) with an addition to the filter.
Suggested replacement:
```py
participation = self.users.filter(virtual=ContestParticipation.LIVE, user=profile).first()
```
Note that `participation` is `None` if the participation does not exist.

File: judge/models/contest.py [hunk 7]
```diff
@@ -182,17 +186,32 @@ def can_see_scoreboard(self, user):
             return False
         if self.hide_scoreboard and not self.is_in_contest(user) and self.end_time > self._now:
             return False
+        if self.partially_hide_scoreboard and not self.is_finished_contest(user):
```
Comment (Ninjaclasher) [inline judge/models/contest.py]: This is 2 queries (one in `can_see_scoreboard` and one in `can_see_full_scoreboard`) per call to `can_see_scoreboard`. Can you reduce it to one?

File: judge/models/contest.py [hunk 8]
```diff
@@ -182,17 +186,32 @@ def can_see_scoreboard(self, user):
...
+            return False
         return True
 
     def can_see_full_scoreboard(self, user):
         if self.is_editable_by(user):
             return True
         if not self.is_accessible_by(user):
+            # Not accessible by user
```
Comment (Ninjaclasher) [inline judge/models/contest.py]: Comment is unnecessary (self explanatory from method name)

File: judge/models/contest.py [hunk 9]
```diff
@@ -259,11 +279,12 @@ def update_user_count(self):
 
     @cached_property
     def show_scoreboard(self):
-        if not self.can_join:
+        # Show scoreboard, contest wise
```
Comment (Ninjaclasher) [inline judge/models/contest.py]: same with these comments

File: judge/models/contest.py [hunk 10]
```diff
@@ -322,7 +343,6 @@ def is_accessible_by(self, user):
                 if user.is_authenticated and not user.profile.is_external_user:
                     return True
 
-        # User can edit the contest
```
Comment (Ninjaclasher) [inline judge/models/contest.py]: ^

**Review by Ninjaclasher — CHANGES_REQUESTED** [review]

File: judge/models/contest.py (hunk 2 above)
Comment (jtyliu) [reply judge/models/contest.py]: Done

File: judge/models/contest.py (hunk 7 above)
Comment (jtyliu) [reply judge/models/contest.py]: I don't see a good way to reduce it down to one query `self.hide_scoreboard` has the same problem.

**Comment by jtyliu** [comment]
Added all changes requested but reducing the two queries down to one. I don't see a good way to do implement it.

File: judge/models/contest.py [hunk 11]
```diff
@@ -191,8 +197,17 @@ def can_see_full_scoreboard(self, user):
...
         if not self.show_scoreboard:
             return False
+        if self.partially_hide_scoreboard and not self.is_finished_contest(user):
+            return False
         return True
 
+    def is_finished_contest(self, user):
+        if user.is_authenticated:
+            participation = self.users.filter(virtual=ContestParticipation.LIVE, user=user.profile).first()
+            if participation and participation.ended and participation.contest == self:
```
Comment (Ninjaclasher) [inline judge/models/contest.py]: No need to check `participation.contest == self`

File: judge/models/contest.py [hunk 12]
```diff
@@ -220,7 +235,6 @@ def contest_window_length(self):
 
     @cached_property
     def _now(self):
-        # This ensures that all methods talk about the same now.
```
Comment (Ninjaclasher) [inline judge/models/contest.py]: ?

File: judge/models/contest.py [hunk 13]
```diff
@@ -81,6 +81,10 @@ class Contest(models.Model):
     registration_end_time = models.DateTimeField(verbose_name=_('registration end time'),
                                                  help_text=_('Allow registration until the specified time.'),
                                                  blank=True, null=True)
+    partially_hide_scoreboard = models.BooleanField(verbose_name=_('Partially hide scoreboard'),
```
Comment (Ninjaclasher) [inline judge/models/contest.py]: Just for consistency, can you move this block to under `hide_scoreboard`?

**Review by Ninjaclasher — COMMENTED** [review]
Otherwise looks good

File: judge/models/contest.py (hunk 12 above)
Comment (jtyliu) [reply judge/models/contest.py]: Um it's there on my side.

**Review by Ninjaclasher — APPROVED** [review]

**Review by fluix-dev — CHANGES_REQUESTED** [review]
LGTM, just needs a migration rebase.

File: judge/migrations/0099_partially_hide_scoreboard.py [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+
+    dependencies = [
+        ('judge', '0098_registration_data'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='contest',
+            name='partially_hide_scoreboard',
+            field=models.BooleanField(default=False, help_text="Whether the scoreboard should be shown remain hidden until the member's contest time is finished", verbose_name='Partially hide scoreboard'),
```
Comment (fluix-dev) [inline judge/migrations/0099_partially_hide_scoreboard.py]:
```suggestion
            field=models.BooleanField(default=False, help_text="Whether the scoreboard should be hidden until the member's contest time is finished", verbose_name='Partially hide scoreboard'),
```

File: judge/models/contest.py [hunk 15]
```diff
@@ -85,6 +85,10 @@ class Contest(models.Model):
                                           help_text=_('Whether the scoreboard should remain hidden for the duration '
                                                       'of the contest.'),
                                           default=False)
+    partially_hide_scoreboard = models.BooleanField(verbose_name=_('Partially hide scoreboard'),
+                                                    help_text=_("Whether the scoreboard should be shown hidden "
```
Comment (fluix-dev) [inline judge/models/contest.py]:
```suggestion
                                                    help_text=_("Whether the scoreboard should be hidden "
```

File: judge/migrations/0099_partially_hide_scoreboard.py [hunk 16]
```diff
@@ -0,0 +1,18 @@
+# Generated by Django 2.2.9 on 2019-12-27 20:27
+
+from django.db import migrations, models
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('judge', '0098_registration_data'),
```
Comment (fluix-dev) [inline judge/migrations/0099_partially_hide_scoreboard.py]:
```suggestion
        ('judge', '0104_contestproblem_maxsubs),
```

**Review by fluix-dev — CHANGES_REQUESTED** [review]

File: judge/migrations/0099_partially_hide_scoreboard.py (hunk 14 above)
Comment (fluix-dev) [reply judge/migrations/0099_partially_hide_scoreboard.py]: Actually this entire `help_text` doesn't match the model's.

Comment (jtyliu) [reply judge/migrations/0099_partially_hide_scoreboard.py]: How about `Whether the scoreboard is hidden until the user finishes their contest window`

Comment (fluix-dev) [reply judge/migrations/0099_partially_hide_scoreboard.py]: Let's go with:
> Whether the scoreboard is hidden until the user's contest window ends

Make sure to change it in the model and the migration.

**Comment by fluix-dev** [comment]
~~Any reason this isn't done the same was as `permanently_hide_scoreboard` by changing [`show_scoreboard`](https://github.com/mcpt/wlmoj/blob/master/judge/models/contest.py#L272) similar to [this](https://github.com/mcpt/wlmoj/blob/master/judge/models/contest.py#L277) and [`_get_queryset`](https://github.com/mcpt/wlmoj/blob/master/judge/views/submission.py#L219) similar to [this](https://github.com/mcpt/wlmoj/blob/master/judge/views/submission.py#L239)?~~

Disregard.

**Review by fluix-dev — APPROVED** [review]

**Review by Ninjaclasher — APPROVED** [review]

**Pull request merged**